File: android/app/src/main/kotlin/com/cruisemesh/app/chat/MeshSender.kt
package com.cruisemesh.app.chat

import android.util.Log
import com.cruisemesh.app.media.AttachmentPayload
import com.cruisemesh.app.media.KIND_ATTACHMENT_MANIFEST
import com.cruisemesh.app.media.KIND_REACTION
import com.cruisemesh.app.mesh.MeshRouter
import com.cruisemesh.app.mesh.RelaySyncEvents
import com.cruisemesh.app.mesh.encodeOutboundEnvelopeFrame
import uniffi.cruisemesh_core.Contact
import uniffi.cruisemesh_core.Identity
import uniffi.cruisemesh_core.MessageStore

private const val TAG = "MeshSender"

/** The `kind` byte for a plaintext chat message (DESIGN.md §7.1). */
private const val KIND_TEXT: kotlin.UByte = 1u

/** Cumulative-receipt type bytes (DESIGN.md §7.2), mirroring the private
 * copies already established in MainActivity.kt / ChatScreen.kt / MeshService.kt. */
/**
 * Sends an outgoing text or media message into a contact's 1:1 chat
 * (DESIGN.md §7.1 / §8). [ChatScreen] depends only on this interface -- never
 * on a concrete implementation -- so the transport can be swapped from a
 * local-only stub to real mesh delivery without any UI changes.
 */
interface MeshSender {
    /** Sends `text` as a `kind=1` message to `contact`'s chat. */
    fun sendText(contact: Contact, text: String, replyToMsgId: ByteArray? = null): SendResult

    /**
     * Sends an inline attachment as a `kind=16` chat-stream message
     * (attachment manifest with embedded blob; DESIGN.md §8).
     */
    fun sendAttachment(
        contact: Contact,
        attachment: AttachmentPayload,
        replyToMsgId: ByteArray? = null,
    ): SendResult

    /** Sends or clears this user's emoji reaction to [target]. */
    fun sendReaction(contact: Contact, target: MessageTarget, emoji: String): SendResult
}

/** Whether an authored message made it into the durable local message/outbound transaction. */
enum class SendResult {
    STORED,
    FAILED,
}

/**
 * Real [MeshSender]: seals the message once into a persistent outbound
 * envelope (DESIGN.md §4, §9), stores both the plaintext row and that exact
 * sealed retry copy atomically in [MessageStore], then transmits it
 * immediately if [MeshRouter] currently has a live link to the contact. If
 * the contact isn't connected right now, the same queued envelope stays local
 * for [com.cruisemesh.app.mesh.MeshService]'s digest sync (and later relay
 * upload) to retry. Because [ChatScreen] only ever depends on the
 * [MeshSender] interface, this swap-in required no UI changes.
 *
 * 1:1 chats use the peer's UserID as the *local* `chat_id` (DESIGN.md §7.1),
 * so the outgoing message is stored under `contact.userId` even though
 * `identity.userId` is the sender. Rust atomically assigns Lamports and
 * ratchets past the peer's delivered/read watermarks.
 *
 * On the *wire*, `MessageBody.chatId` is set to `identity.userId` (our own),
 * not `contact.userId` -- see `MeshService`'s class KDoc for the full
 * explanation of why the wire and local conventions differ.
 */
class RealMeshSender(
    private val store: MessageStore,
    private val identity: Identity,
) : MeshSender {
    override fun sendText(contact: Contact, text: String, replyToMsgId: ByteArray?): SendResult =
        enqueueAuthored(
            contact = contact,
            kind = KIND_TEXT,
            payload = text.toByteArray(Charsets.UTF_8),
            logLabel = "sendText",
            replyToMsgId = replyToMsgId,
        )

    override fun sendAttachment(
        contact: Contact,
        attachment: AttachmentPayload,
        replyToMsgId: ByteArray?,
    ): SendResult {
        if (attachment.blob.size > AttachmentPayload.MAX_BLOB_BYTES) {
            Log.w(TAG, "Refusing attachment larger than ${AttachmentPayload.MAX_BLOB_BYTES} bytes")
            return SendResult.FAILED
        }
        val payload = try {
            attachment.encode()
        } catch (e: Exception) {
            Log.e(TAG, "sendAttachment: could not encode attachment", e)
            return SendResult.FAILED
        }
        return enqueueAuthored(
            contact = contact,
            kind = KIND_ATTACHMENT_MANIFEST,
            payload = payload,
            logLabel = "sendAttachment",
            replyToMsgId = replyToMsgId,
        )
    }

    override fun sendReaction(contact: Contact, target: MessageTarget, emoji: String): SendResult {
        val payload = try {
            ReactionPayload(target, emoji).encode()
        } catch (e: Exception) {
            Log.e(TAG, "sendReaction: could not encode reaction", e)
            return SendResult.FAILED
        }
        return enqueueAuthored(
            contact = contact,
            kind = KIND_REACTION,
            payload = payload,
            logLabel = "sendReaction",
        )
    }

    private fun enqueueAuthored(
        contact: Contact,
        kind: UByte,
        payload: ByteArray,
        logLabel: String,
        replyToMsgId: ByteArray? = null,
    ): SendResult {
        val queued = try {
            val timestamp = System.currentTimeMillis()
            val authored = store.authorPairwiseMessage(identity, contact, kind, payload, replyToMsgId, timestamp)
            // V2 field metric: note the outbound send so its delivery latency
            // and confirmation route can be measured on the cruise test.
            runCatching {
                store.recordSentMetric(authored.message.chatId, authored.message.lamport, timestamp)
            }
            authored.message.chatId to authored.acknowledgedDelivered
        } catch (e: Exception) {
            Log.e(TAG, "$logLabel: message was not stored for ${contact.name}", e)
            return SendResult.FAILED
        }

        val (chatId, ackedDelivered) = queued
        try {
            ChatEvents.notifyChatChanged(chatId)
            RelaySyncEvents.requestSync()

            // Preserve causal order during the one-sided friending window. The
            // friend-card request is an earlier row in this same authored stream.
            // If its first BLE attempt happened before the peer's link was ready,
            // sending only this new text lets the text arrive first as an unknown
            // sender; it then stays hidden until a reverse scan imports us. Replay
            // every still-unacknowledged envelope in lamport order so the card is
            // always queued on the link before the first visible message.
            val pending = store
                .outboundEnvelopesAfter(chatId, identity.userId, ackedDelivered)
                .sortedBy { it.lamport }
            for (pendingEnvelope in pending) {
                val frame = encodeOutboundEnvelopeFrame(pendingEnvelope)
                if (!MeshRouter.sendToUserId(contact.userId, frame)) {
                    // No direct link to the recipient right now -- give the sealed
                    // envelope to whoever IS connected so it can mule to the
                    // recipient later (muling hook A).
                    val muled = MeshRouter.relayToAll(frame)
                    Log.i(
                        TAG,
                        "$logLabel: ${contact.name} not currently connected; sprayed pending lamport=${pendingEnvelope.lamport} to $muled mule link(s)",
                    )
                }
            }
        } catch (e: Exception) {
            // The durable transaction already succeeded. A stale or failing
            // transport must leave the message visible and queued for retry.
            Log.w(TAG, "$logLabel: stored locally; immediate delivery will retry", e)
        }
        return SendResult.STORED
    }
}

File: ios/CruiseMesh/Chat/MeshSender.swift
import Foundation
import os.log

/// Whether an authored message made it into the durable local
/// message/outbound transaction. Mirrors Android's `SendResult`.
///
/// Without it every failure here was a `return` into `Void`, so the composer
/// could not tell "durably queued" from "gone" and cleared the user's typed
/// text either way — the silent send that lost the message.
enum SendResult: Equatable {
    case stored
    case failed
}

protocol MeshSender {
    @discardableResult
    func sendText(contact: Contact, text: String) -> SendResult
    @discardableResult
    func sendAttachment(contact: Contact, attachment: AttachmentPayload) -> SendResult
    @discardableResult
    func sendReaction(contact: Contact, target: MessageTarget, emoji: String) -> SendResult
}

final class RealMeshSender: MeshSender {
    private let store: MessageStore
    private let identity: Identity
    private let log = Logger(subsystem: "com.cruisemesh", category: "MeshSender")

    init(store: MessageStore, identity: Identity) {
        self.store = store
        self.identity = identity
    }

    @discardableResult
    func sendText(contact: Contact, text: String) -> SendResult {
        sendText(contact: contact, text: text, replyToMsgId: nil)
    }

    @discardableResult
    func sendText(contact: Contact, text: String, replyToMsgId: Data?) -> SendResult {
        enqueue(
            contact: contact,
            kind: ProtocolKind.text,
            payload: Data(text.utf8),
            label: "sendText",
            replyToMsgId: replyToMsgId
        )
    }

    @discardableResult
    func sendAttachment(contact: Contact, attachment: AttachmentPayload) -> SendResult {
        sendAttachment(contact: contact, attachment: attachment, replyToMsgId: nil)
    }

    @discardableResult
    func sendAttachment(contact: Contact, attachment: AttachmentPayload, replyToMsgId: Data?) -> SendResult {
        guard attachment.blob.count <= AttachmentPayload.maxBlobBytes else {
            log.warning("Refusing oversized attachment")
            return .failed
        }
        return enqueue(
            contact: contact,
            kind: ProtocolKind.attachmentManifest,
            payload: attachment.encode(),
            label: "sendAttachment",
            replyToMsgId: replyToMsgId
        )
    }

    @discardableResult
    func sendReaction(contact: Contact, target: MessageTarget, emoji: String) -> SendResult {
        guard let payload = ReactionPayload(target: target, emoji: emoji).encode() else { return .failed }
        return enqueue(
            contact: contact,
            kind: ProtocolKind.reaction,
            payload: payload,
            label: "sendReaction"
        )
    }

    @discardableResult
    private func enqueue(
        contact: Contact,
        kind: UInt8,
        payload: Data,
        label: String,
        replyToMsgId: Data? = nil
    ) -> SendResult {
        let timestamp = Int64(Date().timeIntervalSince1970 * 1000)
        let authored: AuthoredEnvelope
        do {
            authored = try store.authorPairwiseMessage(
                identity: identity, contact: contact, kind: kind, payload: payload,
                replyToMsgId: replyToMsgId, timestampMs: timestamp
            )
        } catch {
            // Nothing was stored and nothing was queued: say so, so the
            // composer keeps what the user typed. The core refuses to author
            // for real reasons — a contact whose keys will not seal, a device
            // still being adopted (§9.4) — and each of them used to end here
            // as a swallowed `try?`.
            log.error("\(label, privacy: .public): message was not stored: \(error.localizedDescription, privacy: .public)")
            return .failed
        }
        let chatId = authored.message.chatId
        let delivered = authored.acknowledgedDelivered
        // Everything below is best-effort delivery on top of a transaction that
        // already committed. It must never turn into `.failed`: the message is
        // stored and queued, and a retry would author it a second time.
        //
        // V2 field metric: note the outbound send so its delivery latency and
        // confirmation route can be measured on the cruise test.
        try? store.recordSentMetric(chatId: chatId, lamport: authored.message.lamport, sentAtMs: timestamp)
        ChatEvents.notifyChatChanged(chatId)
        RelaySyncEvents.requestSync()
        // A pending kind-3 friend card must reach the peer before the first
        // visible message, otherwise the message is stored as coming from an
        // unknown sender until a reverse scan. Replay the unacknowledged
        // authored stream in Lamport order on every new send.
        let pending = ((try? store.outboundEnvelopesAfter(
            chatId: chatId,
            senderUserId: identity.userId,
            afterLamport: delivered
        )) ?? []).sorted { $0.lamport < $1.lamport }
        for pendingEnvelope in pending {
            let frame = encodeOutboundEnvelopeFrame(pendingEnvelope)
            if !MeshRouter.sendToUserId(userId: contact.userId, frame: frame) {
                let muled = MeshRouter.relayToAll(frame: frame)
                log.info("\(label, privacy: .public): sprayed pending lamport \(pendingEnvelope.lamport) to \(muled) mule link(s)")
            }
        }
        return .stored
    }
}
